### ISSUE TITLE
More opt-in warnings

### ISSUE THREAD

**Issue by charlesroddie** [fsharp/fslang-suggestions#1197]
I suggest adding a set of optional warnings for use of features which some users might want to discourage in their codebases.

F# has many features, which vary in safety and explicitness, and as it grows features are added which appeal to different groups, and are rarely subtracted. A team may decide that some features are disadvantageous to a codebase, and this currently can only be enforced by a style guide.

I am particularly interested in enabling warnings on use of features that increase informality, decrease explicitness of code, or decrease type safety, but others may have different features that they want to discourage.

**Old features**
- Use of a downcast `:?>`
- Use of a type test `:?`
- Use of reflection (covered in another language suggestion)

**New features**
- Implicit yields

**Proposed features**
- Anonymous type-tagged unions
- Record name punning

Existing warnings are here: https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-options#opt-in-warnings . There are already some that allow control of implicit conversions, which I find very helpful (even to the extent of varying strictness depending on the type of library).

**Comment by ly29** [comment]
Untyped string interpolation while is one I would like to see a warning for.

**Comment by dsyme** [comment]
I'm ok with the community adding any of these, if done to high quality, though "implicit yields" is of no great concern to me. 

Oviously FSharpLint is another option for all these.

There are some others I think that could be added such as 
* more thorough warnings for implicit upcasting 
* more thorough warnings for implicit boxing
* any definition or invocation of a new SRTP constraint
* any definition of a interface with static abstract member

**Comment by cartermp** [comment]
What would be the means of configuring this? I do see this as valuable, but there's no systematic way to identify a language feature outside the bounds of the compiler itself. I could imagine a compiler switch that uses some name/Id that corresponds to [language features in the compiler](https://github.com/dotnet/fsharp/blob/main/src/Compiler/Facilities/LanguageFeatures.fsi), but this doesn't cover older features. That could also then let you configure via a project file.

A sub-feature of this could be to define a set of features (e.g., everything related to unsafe casting and conversions) that are thematically related and let you warn on the whole group with a single gesture.

**Comment by dsyme** [comment]
`--warnon:NNN`

**Comment by vzarytovskii** [comment]
We should also consider looking towards the analyzers SDK which will allow adding such warnings easier.

**Comment by Tarmil** [comment]
These do feel more like lints than compiler warnings to me.

**Comment by cartermp** [comment]
@Tarmil Yeah, I agree. However, to make linters more effective we would likely need to make API additions to FCS so that they can accurately detect and warn on use of these features. That won't help with older features, but certainly anything since F# 5 would have a good way to detect things.

**Comment by ly29** [comment]
Yes you’re right about lint. 

Had forgotten about fsharp lint since I found to be too slow and flakey to be useful… maybe it is useful now.

**Comment by IS4Code** [comment]
I am not a big fan of warnings for "intentional" things. Without context, they don't make any sense:
* Why should the language warn about the use of a particular part of the language? If a feature is undesired, don't have it in the language.
* Why warn the user when he picks a certain feature in the language? If the user didn't want to use it, he wouldn't use it.

This applies to upcast/downcast, however a warning for unintentional use of reflection could be useful, since that could cause issues with trimming etc. The rest is just code style.
